api/src/main/java/org/eclipse/microprofile/config/ConfigValue.java

```diff
@@ -0,0 +1,135 @@
+     * After the time expires the next {@link #get()} will again resolve the value
+     * from the underlying {@link Config}.
+     *
+     * @param value the amount of the TimeUnit to wait
+     * @param timeUnit the TimeUnit for the value
+     *
+     * @return This builder
+     */
+    ConfigValue<T> cacheFor(long value, TimeUnit timeUnit);
+
+    /**
+     * Whether to evaluate variables in configured values.
+     * A variable starts with '${' and ends with '}', e.g.
+     * <pre>
+     * mycompany.some.url=${myserver.host}/some/path
+     * myserver.host=http://localhost:8081
+     * </pre>
+     * If 'evaluateVariables' is enabled, the result for the above key
+     * {@code "mycompany.some.url"} would be:
+     * {@code "http://localhost:8081/some/path"}
+     * @param evaluateVariables whether to evaluate variables in values or not
+     * @return This builder
+     */
+    ConfigValue<T> evaluateVariables(boolean evaluateVariables);
+
+
+    /**
+     * The specified lambda will get called on any config value changes.
+     *
+     * @return This builder
+     */
+    ConfigValue<T> onChange(ConfigChanged valueChange);
+
+    /**
+     * Returns the converted resolved filtered value.
+     * @throws java.util.NoSuchElementException if there is no value present
+     * @return the resolved value
+     */
+    T get();
+
+    /**
+     * Returns the converted resolved filtered value as an Optional.
+     *
+     * @return an {@code Optional} for the given type
+     *
+     * @see #get()
+     */
+    Optional<T> getOptional();
+
+    /**
+     * Resolves the value and split it on each comma (',') character.
+     * If a comma is contained in the values it must get escaped with a preceding backslash (&quot;\,&quot;).
+     * Any backslash needs to get escaped via double-backslash (&quot;\\&quot;).
+     *
+     * @return the list of configured comma separated values or an empty Iterable if no
+     */
+    List<T> getValueList();
+
+    /**
+     * Returns the key given in {@link Config#access(String)}.
+     * @return the original key
+     */
+    String getKey();
+
+    /**
+     * Returns the actual key which led to successful resolution and corresponds to the resolved value.
+     * Otherwise the resolved key should always be equal to the original key.
+     * This method is provided for cases, when arameterized resolution is
+     * requested but the value for such appended key is not found and some of the fallback keys is used.
+     *
+     * This should be called only after calling {@link #get()} otherwise the value is undefined (but likely
+     * null).
+     */
+    String getResolvedKey();
+
+
+    /**
+     * TODO feedback from gunnar: could be interesting to have this functionality also as Config#onChange(ConfigChanged)
+     * Callback which can be used with {@link #onChange(ConfigChanged)}
+     */
+    interface ConfigChanged {
+        <T> void onValueChange(String key, T oldValue, T newValue);
+    }
+}
```

Comment:
@struberg The title of this PR is "ConfigValue implementation for dynamic properties" and regarding this, it doesn't offer any more value than a Provider we already accepted to be used.

The other features are not related much to the title of this PR and I would separate them to other PRs, a PR per feature, because they are mostly orthogonal. Would you agree to split it to more PRs, so that we may merge the migration of TCK tests to arquillian and further discuss the new features added by ConfigValue?